### ISSUE TITLE
Parsing error

### ISSUE THREAD

**Issue by jugmac00** [gocept/gocept.template_rewrite#14]
```Parsing error in BlissCompanyCenter/www/BlissCompanyCenter/employeeEditorInner.xpt:113 
	<a class="btn btn-xs btn-success" tal:attributes="href python: company_center.absolute_url()
                            + '/addAddressForm?addressType=employee&companyId=' + company_ID
                            + '&employeeID=' + employee_ID">
Parsing error in BlissCompanyCenter/www/BlissCompanyCenter/employeeEditorInner.xpt:118 
	<a class="btn btn-xs btn-primary"
                   tal:attributes="href python: company_center.absolute_url()
                            + '/set_employee_address?companyId=' + company_ID
                            + '&employeeId=' + employee_ID">
```

Complete code fragment - which could be used for a test case:
```
            <tal:condition condition="not:address">
              <div class="inlineBlock buttonBox col-md-6 col-xs-12" tal:define="company_ID company/getId;
                        company_center python:here.BlissCompanyCenter; employee_ID employee/getId">
                <a class="btn btn-xs btn-success" tal:attributes="href python: company_center.absolute_url()
                            + '/addAddressForm?addressType=employee&companyId=' + company_ID
                            + '&employeeID=' + employee_ID">
                  Add new address
                </a>
                <a class="btn btn-xs btn-primary"
                   tal:attributes="href python: company_center.absolute_url()
                            + '/set_employee_address?companyId=' + company_ID
                            + '&employeeId=' + employee_ID">Use company's main address</a>
              </div>
            </tal:condition>
```

As a workaround I put the whole code on a super long line:
```
                <a class="btn btn-xs btn-success"
                   tal:attributes="href python: company_center.absolute_url() + '/addAddressForm?addressType=employee&companyId=' + company_ID + '&employeeID=' + employee_ID">
                  Add new address
                </a>
```

... which results in no more parsing error

**Comment by viktordick** [comment]
Your code will not only result in an error during the conversion, but also be invalid in Zope4 IIRC (chameleon is more strict). So you should fix the page template before already.

Instead of putting it in one line, you can also use parens, so

    <span tal:content="python: 'a'
       + 'b'"></span>

is not good, but

    <span tal:content="python:('a'
       + 'b')"></span>

will work.

**Comment by jugmac00** [comment]
Thank you very much! The added parenthesis keeps the code readable and is much better than the super long line.

**Comment by jugmac00** [comment]
Hm, I got some more parsing errors - where adding a parenthesis does not help.

```
<a class="btn btn-xs btn-warning"
   tal:attributes="href python: here.absolute_url() + '/remove_contact_and_redirect?employee_id=' + request.get('employee_id') +'&amp;relation_type=otherMaintenanceManager'">
    Remove contact from contract
</a>
```

Maybe you could explain to me what causes parsing errors so I could fix the remaining errors myself?

**Comment by viktordick** [comment]
I think it's the single ; in this case. A single ; separates different attributes, if you want a semicolon you have to double it.
Are you sure this worked in Zope 2?

**Comment by jugmac00** [comment]
I am not sure what you mean with doubling? Literally doubling the semicolon?

`&amp;;` lead to a template error.

Yes, it worked in Zope 2 - the code is in production for almost 3 years now.

I "fixed" the problem by no longer escaping the ampersand
`&amp;` => `&`

Now it both works and template-rewrite is able to parse the code.

**Comment by jugmac00** [comment]
I have several more parsing errors - most probably related to semicolons.

Could you elaborate on "A single ; separates different attributes" - is there anything documented?

In my codebase there are many `tal:define` s with several variables which get defined in a row, separated by a semicolon. Is this no longer valid?

**Comment by jugmac00** [comment]
One more parsing error - but only after using @viktordick PR #11 - without is gives an AttributeError

I found this little "gem" in a 16 year old dependency:
```
	<td align="left" valign="top" colspan="2"
	  tal:define = "cols options/edit_textarea_cols | request/edit_textarea_cols | python:80;
	  rows options/edit_textarea_rows | request/edit_textarea_rows | python:40;
	  content options/externalfile_content | python:here.getContents(request)">
```

I even did not know pipes are allowed in TAL templates.

I suppose it is once again because of the semicolons, ie multiple statements per tal attribute (define here).

Maybe you could help me one last time:
- doubling the semicolons does not please the template-rewrite parser
- I could either hardcode cols and rows so there is only one dynamic value left
- or I could use cascading define statements to avoid the parse error

But both solutions feel wrong - what would be the correct solution to this problem?

**Comment by viktordick** [comment]
This one is probably due to the spaces around the equal ;)

**Comment by jugmac00** [comment]
Parsed my whole code basis! Thanks a lot once more.

@viktordick So, if above listed **parsing errors** are not bugs in the parser, but probably valid Zope 2 syntax, but invalid Zope 4 / Chameleon syntax, this issue could be closed?

Or what do you think?

**Comment by viktordick** [comment]
Well sometimes it is invalid HTML (like the spaces around the equal sign) which Zope2 was very lenient with. Another problem I frequently had were attributes enclosed by `'` instead of `"` or missing whitespaces between two attributes (`id="test"name="something"`).

Some of the points were, I guess, points where Zope2 is not adhering to its ows specification (like the semicolon in `&amp;`) and Zope4/Chameleon is more strict.

In summary, all these points are things one can fix already in Zope 2.

I guess a PR that extends "Known limitations" in the README might be nice, which would then close this discussion.

**Comment by jugmac00** [comment]
> I guess a PR that extends "Known limitations" in the README might be nice, which would then close this discussion.

@viktordick That's a great idea.

In order to give something back, I'd like to create a PR - I'd postpone it to October, as there will be some more time and of course another **Hacktoberfest** :-) 

https://hacktoberfest.digitalocean.com/

**Comment by jugmac00** [comment]
Before I start working on this issue I will wait until #11 is done.

**Comment by jugmac00** [comment]
@viktordick and @icemac I just stumbled over this issue while finally preparing my migration to Python 3 (tomorrow :dancers: ).

Almost one year has passed and I no longer recall what was planned to add to the known issues section.

Could anybody of you two add it? If not, let's close this issue anyway.

**Comment by icemac** [comment]
Closing the issue as suggested by @jugmac00.
